### ISSUE TITLE
Parent sharing assigns with nested LiveViews via `assign_new/3` does not work when navigating between them

### ISSUE THREAD

**Issue by DaTrader** [phoenixframework/phoenix_live_view#3140]
### Environment

* Elixir version (elixir -v): 1.14.4
* Phoenix version (mix deps): 1.7.11
* Phoenix LiveView version (mix deps): 0.20.9 (and before)
* Operating system: Linux Mint
* Browsers you attempted to reproduce this bug on (the more the merrier):  FF, Chrome
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: Yes

### Actual behavior

The `assign_new/3` used in a nested LiveView to assign a parent LiveView assign works as documented only on a first `mount/3`. All subsequent mounts resulting from navigation simply assign the value provided by the child.

The issue manifests regardless of whether parent LiveView is the same or a different module.

### Expected behavior

A nested LiveView invoking `assign_new/3` should always fetch the parent assign when available.

**Comment by DaTrader** [comment]
```elixir
Application.put_env(:sample, Example.Endpoint,
  http: [ip: {127, 0, 0, 1}, port: 5001],
  server: true,
  live_view: [signing_salt: "aaaaaaaa"],
  secret_key_base: String.duplicate("a", 64)
)

Mix.install([
  {:plug_cowboy, "~> 2.5"},
  {:jason, "~> 1.4"},
  {:gettext, "~> 0.23"},
  {:phoenix, "~> 1.7.11"},
  {:phoenix_ecto, "~> 4.4"},
  {:phoenix_html, "~> 3.3"},
  {:phoenix_live_view, "0.20.9", override: true}
])

defmodule Repro3071.Gettext do
  use Gettext, otp_app: :sample
end

defmodule Example.ErrorView do
  def render(template, _), do: Phoenix.Controller.status_message_from_template(template)
end

defmodule Example.DefaultChild do
  use Phoenix.LiveView, layout: {__MODULE__, :live}
  alias Phoenix.LiveView.JS

  def mount( _params, _session, socket) do
    IO.inspect( connected?( socket), label: "mount default child connected?")

    { :ok, assign_new( socket, :action, fn -> :wrong_default_action end)}
  end

  def render( assigns) do
    ~H"""
    <p>Default child</p>

    <p>Action: <%= @action %></p>

    <br>

    <button type="button" class="bg-blue-200 rounded-[8px] px-[12px] py-[6px]" phx-click={JS.navigate( "/other")}>
      Navigate to other
    </button>
    """
  end
end

defmodule Example.OtherChild do
  use Phoenix.LiveView, layout: {__MODULE__, :live}
  alias Phoenix.LiveView.JS

  def mount( _params, _session, socket) do
    IO.inspect( connected?( socket), label: "mount other child connected?")

    { :ok, assign_new( socket, :action, fn -> :wrong_other_action end)}
  end

  def render( assigns) do
    ~H"""
    <p>Other child</p>

    <p>Action: <%= @action %></p>

    <br>

    <button type="button" class="bg-blue-200 rounded-[8px] px-[12px] py-[6px]" phx-click={JS.navigate("/")}>
      Navigate to default
    </button>
    """
  end
end

defmodule Example.HomeLive do
  use Phoenix.LiveView, layout: {__MODULE__, :live}

  defp phx_vsn, do: Application.spec(:phoenix, :vsn)

  def mount( _params, _session, socket) do
    IO.inspect( connected?( socket), label: "mount parent connected?")

    { :ok, assign( socket, :action, socket.assigns.live_action)}
  end

  def handle_params( _params, _url, socket) do
    { :noreply, socket}
  end

  def child_view( :default), do: Example.DefaultChild
  def child_view( :other), do: Example.OtherChild

  def render("live.html", assigns) do
    ~H"""
    <script src={"https://cdn.jsdelivr.net/npm/phoenix@#{phx_vsn()}/priv/static/phoenix.min.js"}></script>
    <script src={"https://cdn.jsdelivr.net/gh/SteffenDE/phoenix_live_view@feedback_node_added_assets/priv/static/phoenix_live_view.min.js"}></script>
    <script src="https://cdn.tailwindcss.com"></script>
    <script>
      let liveSocket = new window.LiveView.LiveSocket("/live", window.Phoenix.Socket)
      liveSocket.connect()
    </script>
    <style>
      * { font-size: 16px; }
    </style>
    <%= @inner_content %>
    """
  end

  def render( assigns) do
    ~H"""
    <div>
      <%= live_render( @socket, child_view( @live_action), id: "Child-#{ @live_action}") %>
    </div>
    """
  end
end

defmodule Example.Router do
  use Phoenix.Router
  import Phoenix.LiveView.Router

  pipeline :browser do
    plug(:accepts, ["html"])
  end

  scope "/", Example do
    pipe_through(:browser)

    live("/", HomeLive, :default)
    live("/other", HomeLive, :other)
  end
end

defmodule Example.Endpoint do
  use Phoenix.Endpoint, otp_app: :sample
  socket("/live", Phoenix.LiveView.Socket)
  plug(Example.Router)
end

{:ok, _} = Supervisor.start_link([Example.Endpoint], strategy: :one_for_one)
Process.sleep(:infinity)
```

**Comment by josevalim** [comment]
Just one very important note: `assign_new` is a "just in case" optimization, not a guarantee. This sharing is performed Therefore, you must always guarantee the result of the function given to `assign_new/3` is the same as if the value was fetched by the parent. I will add such clear notes to the docs as well.

**Comment by DaTrader** [comment]
> Just one very important note: `assign_new` is a "just in case" optimization, not a guarantee. This sharing is performed Therefore, you must always guarantee the result of the function given to `assign_new/3` is the same as if the value was fetched by the parent. I will add such clear notes to the docs as well.

I don't think I understand. Is `assign_new/3` supposed to be reliably used to share parent assigns on child's `mount/3` or not? If not, what's the alternative to fetching same data twice? Sending it from parent to child via `send/2`?

**Comment by DaTrader** [comment]
Or better question, yet.. why can't it be made to always work?

**Comment by josevalim** [comment]
>  Is assign_new/3 supposed to be reliably used to share parent assigns on child's mount/3 or not? 

As I said, it is not a guarantee, so the answer is no. :)

The issue we need to solve is: we need to share data between two processes. How can we do that efficiently?

1. One option is to have the child LiveView automatically ask the parent. However, what if the parent is busy doing something else? Now we are blocking the child waiting for the parent, so this is a hard choice to do as default. You could do it though if you prefer, `handle_call/3` is available as callback as well.

2. Alternatively, you could send all of the LiveView state from parent to child, but could incur in a lot of copying and unused state.

So what we do instead is that we optimize it when we can. If the child LiveView is mounted with the parent LiveView, we "learn" what are the `assigns_new` in the disconnected render. Then, when you do the connected render, we know what you will need, and we send that upfront. I have clarified these conditions in my commit above.

**Comment by DaTrader** [comment]
Thanks!

**Comment by josevalim** [comment]
If you can isolate the issue, please let us know, you should be able to reproduce in a single file using `IO.puts "hello"`.

**Comment by silverdr** [comment]
It was the situation described in the recent [elixirforum thread](https://elixirforum.com/t/multiple-liveviews-on-the-same-page-sharing-data/70941/17) where I had several LiveView(s) rendered on a single DeadView page, trying to utilise the sharing mechanism, i. e. having the expected assigns available in `conn`, connecting and then executing the queries. That setup is somewhat complex and I worked this currently around by turning the `LiveView`s into `LiveCompnent`s and rendering them out of a single LiveView. It still executes the queries second time (after the first time for putting data into `conn`) thooug. I'll see if I can isolate it into a simple, reproducible example.

**Comment by silverdr** [comment]
@SteffenDE thank you for the explanations. Yes, I see those queries twice and from what you say twice is the best I can get. One question though. When the LiveView process starts "During the disconnected render, each LV uses assign_new to prevent fetching the current_user assign again". So far so good. Now: "the browser connects. The parent LiveView fetches current_user again" [...]

Isn't that LiveView the same process that already has `current_user` available?

**Comment by silverdr** [comment]
Roger, that explains - thank you.

> Furthermore, the disconnected render is just a regular HTTP request. After the server responds to the browser, the process* is gone.

It might have been ignorance but I thought/hoped that the process would remain active and await (possibly with timeout) the connection after the initial render. That would allow it not to redo some (potentially expensive) steps like unneeded (in such case) DB roundtrips.

If that's not the case then it explains, even if those double (at least) queries don't feel "right".

**Comment by SteffenDE** [comment]
We've been thinking about ways to maybe make something like that work in the future (see also https://github.com/phoenixframework/phoenix_live_view/issues/3357), but it is complicated. One thing to keep in mind is that for LiveView applications where the majority of views are LiveViews and you're navigating in the same live session, this "double fetching" only occurs on the initial navigation. So most of the time it is not as expensive as it may seem at first.

**Comment by silverdr** [comment]
Right, in an all LiveView application, while startup time is also important, statistically it is probably rare to do lots of costly operations on the entry screen. All in all I am happy to hear that something like I imagined it would/could work is being considered! Thank you once more for thorough explanations.

**Comment by DaTrader** [comment]
@silverdr Have you considered not fetching anything for the static render? 

It depends on the app, but in what I've been building static renders are virtually useless in terms of interaction so I don't even render the components until LiveView is connected.

**Comment by DaTrader** [comment]
@silverdr Yes. You're free to choose what if anything at all you're going to load and or render in the disconnected state. This will depend on your app, but in mine I render just the surrounding components such as the sidebar menu, the header if any and choose not to fetch anything (unless I use cache in which case it's virtually free to fetch twice). 

Since it's my LiveComponents that do all the loading and contain all the app logic I choose not to render them unless `@connected?` (assigned in the LiveView's mount.) It's easier and less error prone to write LiveComponent logic if they ever only run when connected plus I can subscribe, register, send messages, etc. none of which makes any sense in the disconnected state, if for no other reason then because it's all going to get undone once restarted anew when connected.

In short, you need to ask yourself if and why you need that static render stage to fetch/render anything at all.  Do you want the app to provide some basic functionality even if the browser has JS disabled? If not, you don't really need to do anything while not connected.
